testing/e2e-testing.md

```diff
@@ -0,0 +1,100 @@
+# End-to-end testing for Resource Topology Aware Scheduling
+
+A number of end-to-end tests are needed in order to ensure first that the Resource Topology Aware Scheduling solution works as intended, and secondly that the solution has no unintended side effects on a Kubernetes cluster.
+
+The Resource Topology Aware Scheduling system is made up of five active components from a testing perspective: Kubelet, Topology Exporter, Topology Scheduler, Kubernetes API server and the  etcd database. 
+
+![Components in the Resource Topology Scheduling system](images/component_topology.png)
+
+|#  	| Communication		| Motivation				|
+|---	|---			|---					|	
+|1 	| Exporter -> Kubelet		| Exporter talks to Kubelet to read information about Resource state				|
+|2	| Exporter -> API Server	| Exporter talks to the API server to update ResourceTopology Objects				|
+|3 	| Kubelet -> API Server		| Kubelet talks to API server to update node info and assign unbound Pods			|
+|4	| Topol Scheduler -> API Server	| Topology Scheduler plugin talks to the API server looking for NodeResourceTopology objects	|
+|5 	| Scheduler -> API Server	| Scheduler talks to the API server looking for unassigned Pods and Node objects		|
+|6	| API Server -> etcd	| API server persists cluster objects to etcd 							| 
+ 
+
+Three components require changes to an existing Kubernetes cluster as of Kubernetes 1.20:
+
+- **Kubelet** requires changes to expose additional info through the PodResources API.
+- **Exporter** is a new component which creates and updates an API object (NodeResourceTopology). This component may in future be part of NPD.
+- **Scheduler** a plugin for the Kubernetes scheduler will make scheduling decisions based on NodeResourceTopology. This change may be in tree as part of the core Kubernetes scheduler repo.
+
+## Functional testing 	
+Functional testing should show that the Topology Aware Scheduling system performs to spec. Functional testing does not need to be carried out at scale, but should be shown to work on a high-availability cluster with a double-digit number of nodes. 
+
+### Points of failure
+In a multicomponent system like Topology Aware Scheduling there are a large number of points at which failure might occur. Each of the points below represents a class of errors that could be triggered in a large number of scenarios. Below is a list of areas that will be the focus of functional testing:
+
+
+1) Scheduler side effects. Topology Aware Scheduling should have no impact on scheduling of workloads which don't require topology aware scheduling.
+2) Out of date information. There may be drift between the topology crd and nodeInfo. 
```

Comment:
Further, this problem will be solved by adding "hint" into pod's annotation at scheduling stage. This "hint" will be used by node's component to align resources according to "hint". If it will be done scheduler will be able to make more correct assumption (like it assuming now by NodeInfo.Requested).

Reply:
This should be enough to prevent the issue  - I've been using the schedulerName field along with a policy plugin instead of an annotation in order to do this. I think this is the "native" approach, but it's more limited i.e. a single string vs. annotation which could be a more complex object.

Reply:
that's a nice idea and we can implement it already with tools we have (example: https://github.com/fromanirh/k8s-device-plugins - numacell plugin will force topology manager to either pick the hinted numa cell or fail). The problem seems that one way or the other, we need to expose all along the chain the numa cell ID the scheduler wants the pods to run into, something that reportedly violates the amount of information about the HW that k8s seems willing to expose. IOW, we need to expose a too low level detail